components/collections/codepointtrie_builder/cpp/list_to_ucptrie.cpp

```diff
@@ -2,111 +2,85 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
-#include "unicode/umutablecptrie.h"
-#include "unicode/errorcode.h"
-#include "writesrc.h"
+#include <string.h>
 
 #include <iostream>
 
-inline std::int32_t read_int32_t() {
-    std::int32_t x(0);
-    char ch;
-	int sgn = 1;
-	while (!isdigit(ch = getchar_unlocked())) {
-		if (ch == '-') {
-			sgn *= -1;
-		}
-	}
-	x = ch - '0';
-	while (isdigit(ch = getchar_unlocked())) {
-		x = x * 10 + (ch - '0');
-	}
-	return x * sgn;
-}
+#include "unicode/errorcode.h"
+#include "unicode/umutablecptrie.h"
+#include "writesrc.h"
 
 /**
  * list_to_ucptrie: converts a stream of integers to a UCPTrie
  *
- * - standard input should be a stream of integers, each integer representing the value for
-     consecutive code points, starting from code point 0.
+ * - standard input should be a stream of integers, each integer representing
+ the value for consecutive code points, starting from code point 0.
  * - standard output is the build UCPTrie as a TOML file
  */
-int main(int argc, char const *argv[]) {
-    if (argc != 5) {
-        std::cerr << "Takes 4 positional arguments: default value, error value, trie type, and value width" << std::endl;
-        return 1;
-    }
-
-    int32_t defaultValue = atoi(argv[1]);
-    int32_t errorValue = atoi(argv[2]);
 
-    UCPTrieType trieType;
-    std::string trieTypeStr(argv[3]);
-    if (trieTypeStr == "small") {
-        trieType = UCPTRIE_TYPE_SMALL;
-    } else if (trieTypeStr == "fast") {
-        trieType = UCPTRIE_TYPE_FAST;
-    } else {
-        std::cerr << "Expected 'small' or 'fast' for 3rd argument" << std::endl;
-        return 1;
-    }
+constexpr char TRIE_SMALL[] = "small";
+constexpr char TRIE_FAST[] = "fast";
+extern "C" {
+int construct_ucptrie(const int32_t defaultValue, const int32_t errorValue,
+                      const int32_t trieTypeInt, const int32_t valueWidthInt,
+                      const uint32_t* values, const uint32_t n) {
+  UCPTrieType trieType;
+  if (trieTypeInt == 0) {
+    trieType = UCPTRIE_TYPE_SMALL;
+  } else if (trieTypeInt == 1) {
+    trieType = UCPTRIE_TYPE_FAST;
+  } else {
+    fprintf(stderr,
+            "Expected '0' for small or '1' for fast as 3rd argument. Got %d",
+            trieTypeInt);
+    return 1;
+  }
 
-    UCPTrieValueWidth valueWidth;
-    int32_t valueWidthInt = atoi(argv[4]);
-    if (valueWidthInt == 8) {
-        valueWidth = UCPTRIE_VALUE_BITS_8;
-    } else if (valueWidthInt == 16) {
-        valueWidth = UCPTRIE_VALUE_BITS_16;
-    } else if (valueWidthInt == 32) {
-        valueWidth = UCPTRIE_VALUE_BITS_32;
-    } else {
-        std::cerr << "Expected '8', '16', or '32' for 4th argument" << std::endl;
-        return 1;
-    }
+  UCPTrieValueWidth valueWidth;
+  if (valueWidthInt == 8) {
```

Comment:
thought: you could do this on the Rust side. The logic is already there (https://github.com/unicode-org/icu4x/blob/b149924e9c196b44f079c764e83f86b92e477372/components/collections/codepointtrie_builder/src/native.rs#L114-L119), you could make fetching the ICU4C value a method on CPTB. Also for the type.

Reply:
We already do this in native.rs. Moved that logic to CPTB for trie_type and widht.